### PR TITLE
Export batch generated weekly playlists to Spotify

### PR DESCRIPTION
**Pull request by amCap1712**
Add support for exporting the batch generated playlists automatically to Spotify. Note that ListenBrainz stores the latest 2 versions of the playlists whereas in Spotify, we create only one playlist and that reflects the latest version of the LB playlist.

### CODE REVIEW

File: docker/services/cron/crontab [hunk 1]
```diff
@@ -25,8 +25,8 @@ MAILTO=""
 2 2 * * 1 root /usr/local/bin/python /code/listenbrainz/manage.py spark request_missing_mb_data >> /logs/stats.log 2>&1
 # Calculate user similarity
 3 2 * * * root /usr/local/bin/python /code/listenbrainz/manage.py spark cron_request_similar_users >> /logs/stats.log 2>&1
-4 2 * * * root /usr/local/bin/python /code/listenbrainz/manage.py spark request_troi_playlists --slug weekly-jams >> /logs/troi_spark.log
-5 2 * * * root /usr/local/bin/python /code/listenbrainz/manage.py spark request_troi_playlists --slug weekly-exploration >> /logs/troi_spark.log
+4 2 * * 0 root /usr/local/bin/python /code/listenbrainz/manage.py spark request_troi_playlists --slug weekly-jams >> /logs/troi_spark.log
```
Comment (mayhem) [inline docker/services/cron/crontab:28]: I wonder if we should hold off on making this change -- now we need to ensure that all the new code for sending to spotify works, and doing it on a daily basis for now will let us solve this much faster

File: listenbrainz/troi/weekly_playlists.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    template = SQL("""({creator_id}, %s, {description}, 't', %s, %s)""").format(
+        creator_id=Literal(TROI_BOT_USER_ID),
+        description=Literal(description)
+    )
+    values = [(p["name"], p["user_id"], json.dumps(p["additional_metadata"])) for p in playlists]
+    results = execute_values(cursor, query, values, template, fetch=True)
+    return {r[0]: r[1] for r in results}
+
+
+def filter_and_update_playlists(slug, jam_name, all_playlists):
```
Comment (mayhem) [inline listenbrainz/troi/weekly_playlists.py:102]: It took me quite a while to work out that the playlist are not being deleted anywhere, but are being removed from the all_playlists list. Perhaps this function should be renamed? exclude_playlists_from_deleted_users?

**Review by mayhem — APPROVED** [review]
Some minor improvements, free to merge after.

File: listenbrainz/troi/weekly_playlists.py (hunk 2 above)
Comment (amCap1712) [reply listenbrainz/troi/weekly_playlists.py:102]: It excludes playlists and also updates the metadata for the remaining but I guess I could split it into two.

Comment (mayhem) [reply listenbrainz/troi/weekly_playlists.py:102]: I didn't mean for it to be split, just change the name to be more representative and all good.

Comment (amCap1712) [reply listenbrainz/troi/weekly_playlists.py]: Makes sense

File: docker/services/cron/crontab (hunk 1 above)
Comment (amCap1712) [reply docker/services/cron/crontab]: Reverted.

File: listenbrainz/troi/weekly_playlists.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            playlist_element = Playlist(name=playlist["name"], description=description, recordings=recordings)
+            playlist_url, _ = submit_to_spotify(sp, playlist_element, user["external_user_id"],
+                                                existing_url=playlist["existing_url"])
+            playlist["additional_metadata"].update({"external_urls": {"spotify": playlist_url}})
+        except Exception:
+            current_app.logger.error("Unable to export playlist to spotify:", exc_info=True)
+
+
+def insert_recordings(cursor, playlist_id: str, recordings: list):
+    query = "INSERT INTO playlist.playlist_recording (playlist_id, position, mbid, added_by_id) VALUES %s"
```
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/troi/weekly_playlists.py:79]: done

**Pull request merged**